File: Data Structures & Algorithms/binary-tree-diameter/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        # involves depth of tree
        if not root:
            return 0
        return max(self.diameterOfBinaryTree(root.left), self.diameterOfBinaryTree(root.right), self.max_depth(root.left) + self.max_depth(root.right))

    def max_depth(self, root) -> int:
        if not root:
            return 0
        return 1 + max(self.max_depth(root.left), self.max_depth(root.right))

File: Data Structures & Algorithms/balanced-binary-tree/submission-0.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} root
     * @return {boolean}
     */
    isBalanced(root) {
        if(!root) return true;

        return this.isBalanced(root.left) && this.isBalanced(root.right) && Math.abs(this.maxHeight(root.left) - this.maxHeight(root.right)) <= 1;
    }

    maxHeight(root) {
        if(!root) return 1;

        return 1 + Math.max(this.maxHeight(root.left), this.maxHeight(root.right))
    }
}

/*

find max height of left and right, see if they differ by 1+

*/